test/abc/sound/SequencePlayerTest.java

```diff
@@ -1,38 +1,125 @@
+			midiUe.printStackTrace();
+		} catch (InvalidMidiDataException invalidMidiDe) {
+			invalidMidiDe.printStackTrace();
+		} catch (InterruptedException intEx) {
+			intEx.printStackTrace();
+		}
+	}
+
+	@Test
+	public void piece2() {
 		try {
-			CharStream input = fromFileName(fileName);
-			AbcLexer lexer = new AbcLexer(input);
-			CommonTokenStream tokens = new CommonTokenStream(lexer);
-			AbcParser parser = new AbcParser(tokens);
-			ParseTree tree = parser.abc_tune();
-			// TODO: Implement custom walker, parse tree into POJO, play notes using SquencePlayer
-			ParseTreeWalker walker = new ParseTreeWalker();
-			AbcBaseListener listener = new AbcBaseListener();
-			walker.walk(listener, tree);
-		} catch (IOException e) {
-			e.printStackTrace();
+			SequencePlayer player = new SequencePlayer(200, 12);
+
+			player.addNote(new Pitch('E').transpose(1).toMidiNote(), 0, 6);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 0, 6);
+			player.addNote(new Pitch('E').transpose(1).toMidiNote(), 6, 6);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 6, 6);
+
+			player.addNote(new Pitch('E').transpose(1).toMidiNote(), 18, 6);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 18, 6);
+
+			player.addNote(new Pitch('C').transpose(1).toMidiNote(), 30, 6);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 30, 6);
+
+			player.addNote(new Pitch('E').transpose(1).toMidiNote(), 36, 12);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 36, 12);
+
+			player.addNote(new Pitch('G').toMidiNote(), 48, 12);
+			player.addNote(new Pitch('B').toMidiNote(), 48, 12);
+			player.addNote(new Pitch('G').transpose(12).toMidiNote(), 48, 12);
+
+			player.addNote(new Pitch('G').toMidiNote(), 72, 12);
+
+			player.addNote(new Pitch('C').transpose(12).toMidiNote(), 96, 18);
+			player.addNote(new Pitch('G').toMidiNote(), 114, 6);
+
+			player.addNote(new Pitch('E').toMidiNote(), 132, 12);
+
+			player.addNote(new Pitch('E').toMidiNote(), 144, 6);
+			player.addNote(new Pitch('A').toMidiNote(), 150, 12);
+			player.addNote(new Pitch('B').toMidiNote(), 162, 12);
+			player.addNote(new Pitch('B').transpose(-1).toMidiNote(), 174, 6);
+			player.addNote(new Pitch('A').toMidiNote(), 180, 12);
+
+			player.addNote(new Pitch('G').toMidiNote(), 192, 8);
+			player.addNote(new Pitch('E').transpose(12).toMidiNote(), 200, 8);
+			player.addNote(new Pitch('G').transpose(12).toMidiNote(), 208, 8);
+
+			player.addNote(new Pitch('A').transpose(12).toMidiNote(), 216, 12);
+			player.addNote(new Pitch('F').transpose(12).toMidiNote(), 228, 6);
+			player.addNote(new Pitch('G').transpose(12).toMidiNote(), 234, 6);
+
+			player.addNote(new Pitch('E').transpose(12).toMidiNote(), 246, 12);
+			player.addNote(new Pitch('C').transpose(12).toMidiNote(), 258, 6);
+			player.addNote(new Pitch('D').transpose(12).toMidiNote(), 264, 6);
+			player.addNote(new Pitch('B').toMidiNote(), 270, 9);
+
```

Comment:
I noticed you are missing the rests `'z'` notes. If you put a Pitch value of '0' it is a rest. Could you please add them in?

Reply:
yup, sure!